vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/UriTemplateImpl.java

```diff
@@ -0,0 +1,375 @@
+  }
+
+  public static final class Literals extends Term {
+    private final String value;
+    private Literals(String value) {
+      this.value = value;
+    }
+  }
+
+  public static final class Expression extends Term {
+    private final Operator operator;
+    private final List<Varspec> value = new ArrayList<>();
+    public Expression(Operator operator) {
+      this.operator = operator;
+    }
+  }
+
+  public static final class Varspec {
+    private final String varname;
+    private Varspec(String varname) {
+      this.varname = varname;
+    }
+  }
+
+  @Override
+  public String expand(MultiMap variables) {
+    StringBuilder sb = new StringBuilder();
+    terms.forEach(term -> {
+      if (term instanceof Literals) {
+        sb.append(((Literals)term).value);
+      } else {
+        Expression expression = (Expression) term;
+        expression.operator.expand(expression.value, variables, sb);
+      }
+    });
+    return sb.toString();
+  }
+
+  public static class Parser {
+
+    private UriTemplateImpl template;
+    private Expression expression;
+
+    public UriTemplateImpl parseURITemplate(String s) {
+      template = new UriTemplateImpl();
+      if (parseURITemplate(s, 0) != s.length()) {
+        throw new IllegalArgumentException();
+      }
+      return template;
+    }
+
+    public int parseURITemplate(String s, int pos) {
+      while (true) {
+        int idx = parseLiterals(s, pos);
+        if (idx > pos) {
+          template.terms.add(new Literals(s.substring(pos, idx)));
+          pos = idx;
+        } else {
+          idx = parseExpression(s, pos);
+          if (idx > pos) {
+            pos = idx;
+          } else {
+            break;
+          }
+        }
+      }
+      return pos;
+    }
+
+    public int parseExpression(String s, int pos) {
+      if (pos < s.length() && s.charAt(pos) == '{') {
+        int idx = pos + 1;
+        Operator operator;
+        if (idx < s.length() && isOperator(s.charAt(idx))) {
+          operator = mapping.get(s.charAt(idx));
+          idx++;
+        } else {
+          operator = Operator.SIMPLE_STRING_EXPANSION;
+        }
+        expression = new Expression(operator);
+        idx = parseVariableList(s, idx);
+        if (idx < s.length() && s.charAt(idx) == '}') {
+          pos = idx + 1;
+        }
+        if (template != null) {
+          template.terms.add(expression);
+        }
+        expression = null;
+      }
+      return pos;
+    }
+
+    private static boolean isALPHA(int cp) {
+      return ('A' <= cp && cp <= 'Z')
+        || ('a'<= cp && cp <= 'z');
+    }
+
+    public static int parseDIGIT(String s, int pos) {
+      if (pos < s.length() && isDIGIT(s.charAt(pos))) {
+        pos++;
+      }
+      return pos;
+    }
+
+    private static boolean isDIGIT(int cp) {
+      return ('0' <= cp && cp <= '9');
+    }
+
+    private static boolean isHEXDIG(int cp) {
+      return isDIGIT(cp) || ('A' <= cp && cp <= 'F');
```

Comment:
Given that this is the URL alphabet, case should not matter, right? `%0a` should be the same as `%0A`

Reply:
that's correct, the https://datatracker.ietf.org/doc/html/rfc6570 spec says:

```
HEXDIG         =  DIGIT / "A" / "B" / "C" / "D" / "E" / "F"
                       ; case-insensitive
```

I ignored the comment, I'll change that